components/payments/cmd/connectors/internal/storage/balances.go

```diff
@@ -2,59 +2,119 @@ package storage
 
 import (
 	"context"
+	"database/sql"
+	"errors"
 
 	"github.com/formancehq/payments/internal/models"
+	"github.com/uptrace/bun"
 )
 
 func (s *Storage) InsertBalances(ctx context.Context, balances []*models.Balance, checkIfAccountExists bool) error {
 	if len(balances) == 0 {
 		return nil
 	}
 
-	query := s.db.NewInsert().
-		Model((*models.Balance)(nil)).
-		With("cte1", s.db.NewValues(&balances)).
-		Column(
-			"created_at",
-			"account_id",
-			"balance",
-			"currency",
-			"last_updated_at",
-		)
-	if checkIfAccountExists {
-		query = query.TableExpr(`
-		(SELECT *
-			FROM cte1
-			WHERE EXISTS (SELECT 1 FROM accounts.account WHERE id = cte1.account_id)
-				AND cte1.balance != COALESCE((SELECT balance FROM accounts.balances WHERE account_id = cte1.account_id AND last_updated_at < cte1.last_updated_at AND currency = cte1.currency ORDER BY last_updated_at DESC LIMIT 1), cte1.balance+1)
-		) data`)
-	} else {
-		query = query.TableExpr(`
-		(SELECT *
-			FROM cte1
-			WHERE cte1.balance != COALESCE((SELECT balance FROM accounts.balances WHERE account_id = cte1.account_id AND last_updated_at < cte1.last_updated_at AND currency = cte1.currency ORDER BY last_updated_at DESC LIMIT 1), cte1.balance+1)
-		) data`)
+	tx, err := s.db.BeginTx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+
+	for _, balance := range balances {
+		if err := s.insertBalances(ctx, tx, balance, checkIfAccountExists); err != nil {
+			return err
+		}
 	}
 
-	query = query.On("CONFLICT (account_id, created_at, currency) DO NOTHING")
+	return e("failed to commit transaction", tx.Commit())
+}
 
-	_, err := query.Exec(ctx)
+func (s *Storage) insertBalances(ctx context.Context, tx bun.Tx, balance *models.Balance, checkIfAccountExists bool) error {
+	var account models.Account
+	err := tx.NewSelect().
+		Model(&account).
+		Where("id = ?", balance.AccountID).
+		Scan(ctx)
```

Comment:
Shouldn't we skip running this query at all if `checkIfAccountExists` is false? Or else, what is the purpose of this flag?

Previously it was only checking if the account existed in `accounts.account` if this flag is true.

Reply:
The purpose of this flag is to not insert the balance if the account is not existing, so we need this check before I think

Reply:
It doesn't look like that to me.

The crucial difference between the two SQL statements is whether `WHERE EXISTS (SELECT 1 FROM accounts.account WHERE id = cte1.account_id)` is run or not.
So if `checkIfAccountExists` is true, and the above SQL doesn't return results, then the balance insertion will be skipped.

However if `checkIfAccountExists` is false it doesn't care if there is a corresponding entry in `accounts.account` or not, and will proceed with the balance update as long as it can find entries that match `balance.account_id` that meet the criteria.

I have to admit I don't fully understand the use-case though. In what cases do we have `accounts.balance.account_id` that don't correspond to records in `accounts.account` ?

Reply:
Since there is a foreign key on the db, I understand now what you mean about how there would have been an error in the original implementation. That resolves my comment.